link/kprobe.go

```diff
@@ -0,0 +1,231 @@
+// number of programs.
+func Kprobe(symbol string) (*traceEvent, error) {
+	return kprobe(symbol, false)
+}
+
+// Kretprobe returns a TraceEvent that fires right before the given kernel
+// symbol exits, with the function stack left intact.
+// See /proc/kallsyms for available symbols. For example, printk():
+//
+//	Kretprobe("printk")
+//
+// As of kernel 4.15, the presence of the given symbol is only checked when
+// attaching a program. One TraceEvent can be re-used to attach an arbitrary
+// number of programs.
+func Kretprobe(symbol string) (*traceEvent, error) {
+	return kprobe(symbol, true)
+}
+
+// kprobe returns a TraceEvent backed by a k(ret)probe
+// of the given kernel symbol.
+func kprobe(symbol string, ret bool) (*traceEvent, error) {
+	if symbol == "" {
+		return nil, errors.New("symbol name cannot be empty")
+	}
+	if !rgxTraceEvent.MatchString(symbol) {
+		return nil, fmt.Errorf("symbol must be alphanumeric (with _ and -): %s", symbol)
+	}
+
+	// Use kprobe PMU if the kernel has it available.
+	tp, err := pmuKprobe(symbol, ret)
+	if err == nil {
+		return tp, nil
+	}
+	if err != nil && !errors.Is(err, ErrNotSupported) {
+		return nil, fmt.Errorf("creating perf_kprobe PMU: %w", err)
+	}
+
+	// Use tracefs if kprobe PMU is missing.
+	tp, err = tracefsKprobe(traceGroup, symbol, ret)
+	if err != nil {
+		return nil, fmt.Errorf("creating trace event '%s' in tracefs: %w", symbol, err)
+	}
+
+	return tp, nil
+}
+
+// pmuKprobe returns a TraceEvent to be created using PMU (sysfs).
+func pmuKprobe(symbol string, ret bool) (*traceEvent, error) {
+	et, err := getPMUEventType("kprobe")
+	if err != nil {
+		return nil, err
+	}
+	return &traceEvent{
+		pmuID:     et,
+		name:      symbol,
+		ret:       ret,
+		progType:  ebpf.Kprobe,
+		finalizer: nil, // no resources are created
+	}, nil
+}
+
+// tracefsKprobe returns a TraceEvent created by writing an entry to
+// <tracefs>/kprobe_events. On Close(), the entry is removed.
+// Repeated calls with the same arguments will yield TraceEvents without
+// finalizers, making sure only one removal attempt against
+// <tracefs>/kprobe_events is made during program teardown.
+func tracefsKprobe(group, symbol string, ret bool) (*traceEvent, error) {
+	te := traceEvent{
+		group:     group,
+		name:      symbol,
+		ret:       ret,
+		progType:  ebpf.Kprobe, // kernel only allows kprobe programs to attach
+		finalizer: nil,
+	}
+
+	// Before attempting to create a trace event through tracefs,
+	// check if an event with the same name already exists.
+	tid, err := getTraceEventID(group, kprobeTraceEventName(symbol, ret))
+	if err != nil && !errors.Is(err, ErrNotSupported) {
+		return nil, fmt.Errorf("getting trace event id: %w", err)
+	}
+	if err == nil {
+		// Kprobe trace event already exists, return a TraceEvent to the caller.
+		// Since this kprobe event was created by a prior call to this function,
+		// or even by another tool or process, it should not be removed by Close().
```

Comment:
Correct me if I am wrong, another tool or process shouldn't manage probes in the same group/namespace of this instance, right?
If we start using this probe and then the other process deletes it, the probe will "stop working" but for the library everything is ok.

In theory the same thing can happen in the same instance:

`a, _ = Kprobe("__x64_sys_execve")`
`b, _ = Kprobe("__x64_sys_execve")`

What about instanciating probes with an identifier? Example:

`p:ebpf_1234/p_my_kprobe_123 __x64_sys_execve`
`p:ebpf_1234/p_my_kprobe_456 __x64_sys_execve`

This way we should always expect the finalizer to be present (and should then be safe to call it)

Reply:
True, I've given it some thought, and not taking ownership in some cases will always be brittle. Will generate a random group name for each creation instead, which means the finalizer system can be removed completely. This will allow `TraceEvent`s to be serialized using `gob` for tools that want to implement seamless restarts and will still allow those tools to clean up after themselves properly.